### ISSUE TITLE
Tracking Issue: More core blocks and new `store()` API

### ISSUE THREAD

**Issue by luisherranz** [WordPress/gutenberg#53740]
- Part of #41236.

In [Expose the full Interactivity API in `@wordpress/interactivity`](https://github.com/WordPress/gutenberg/issues/51056), we exposed a publicly testable version of the Interactivity API so people can start experimenting with it.

For the next iteration of the Interactivity API, we want to focus on four projects, including 3 core blocks (Comments Form, Query and Search blocks) and a major overhaul of the Store API. We'll update the list of tasks as we progress on the different projects.

- **Query block client-side pagination** ([#53812](https://github.com/WordPress/gutenberg/pull/53812))
  - [x] Directive to do client-side navigation of the inner contents of a block
    - https://github.com/WordPress/gutenberg/pull/53733
  - [x] Add support for a `data-wp-key` directive that is linked to Preact's `key`
    - https://github.com/WordPress/gutenberg/pull/53844  
  - [x] Merge new SSR context on client-side navigation
    - https://github.com/WordPress/gutenberg/pull/53853
  - [x] Cancel navigation when another starts
    - https://github.com/WordPress/gutenberg/discussions/53065
    - https://github.com/WordPress/gutenberg/pull/54201
  - [x] Add timeout to client navigation
    - https://github.com/WordPress/gutenberg/pull/54474
  - [x] Add e2e tests for fallback to server navigation when the client-side navigation fails
    - https://github.com/WordPress/gutenberg/discussions/53064
  - [x] Add support for `data-wp-fill`/`data-wp-slot` directives (without SSR support yet)
    - https://github.com/WordPress/gutenberg/pull/53958 
  - [x] Fix design/UX issues
    - https://github.com/WordPress/gutenberg/issues/54059 
    - https://github.com/WordPress/gutenberg/pull/54347
    - https://github.com/WordPress/gutenberg/issues/54414
  - [x] Fix accessibility issues
    - Query block: announce dynamically loaded content ([conversation](https://github.com/WordPress/gutenberg/pull/53812#discussion_r1307601024))
    - Query block: maintain focus on clicked link when navigating ([conversation](https://github.com/WordPress/gutenberg/pull/53812#discussion_r1307647335))
  - [x] Update the `<title>` on navigation.
    - https://github.com/WordPress/gutenberg/pull/55446
  - [x] Turn off the enhanced pagination when certain blocks are inserted: Post Content, Template Part, Reusable block…
    - https://github.com/WordPress/gutenberg/pull/55539
- **New store() API**:
  - [x] Make the modifications in Woo.
    - https://github.com/woocommerce/woocommerce-blocks/issues/11065
    - https://github.com/woocommerce/woocommerce-blocks/pull/11071
  - [x] Open a PR with the new API here in Gutenberg to make sure that the e2e tests that we have in Gutenberg pass.
    - https://github.com/WordPress/gutenberg/pull/55459 
  - [x] Migrate the Woo blocks to the new API to make sure that everything works fine and we don't find any blockers.
    - https://github.com/woocommerce/woocommerce-blocks/pull/11558
    - https://github.com/woocommerce/woocommerce-blocks/pull/11721
  - [x] Sync the code back to Gutenberg.
    - https://github.com/WordPress/gutenberg/pull/55459 
  - [x] Migrate the Core blocks to the new API to make sure that everything works fine and we don't fine any blockers.
    - https://github.com/WordPress/gutenberg/pull/55459 
  - [x] Migrate the [create-block template](https://github.com/WordPress/gutenberg/tree/trunk/packages/create-block-interactive-template)
    - https://github.com/WordPress/gutenberg/pull/56613
  - [x] #56640
    - https://github.com/WordPress/gutenberg/pull/56764
  - [x] Migrate [proposal examples](https://make.wordpress.org/core/2023/03/30/proposal-the-interactivity-api-a-better-developer-experience-in-building-interactive-blocks/).
- **Others**:
  - [x] https://github.com/WordPress/gutenberg/issues/56641
    - https://github.com/WordPress/gutenberg/pull/56764
  - [x] Migration guide
    - https://github.com/WordPress/gutenberg/discussions/52906#discussioncomment-7810998

---

Please subscribe to keep track of the progress.

**Comment by luisherranz** [comment]
I've opened PRs for the enhanced ux of the Query block and Comments block:

- https://github.com/WordPress/gutenberg/pull/53737
- https://github.com/WordPress/gutenberg/pull/53812

We'll work on them as we progress through these tasks.

**Comment by luisherranz** [comment]
I've opened a PR to add support for `data-wp-key`: 

- https://github.com/WordPress/gutenberg/pull/53844

**Comment by luisherranz** [comment]
I've added "Merge new SSR context on client-side navigation" because we need it for the Query block. I'll start working on it:

- https://github.com/WordPress/gutenberg/discussions/53057

**Comment by luisherranz** [comment]
I've opened a PR to address the previous task:

- https://github.com/WordPress/gutenberg/pull/53853

**Comment by luisherranz** [comment]
I had to add a small fix for the previous PR:

- https://github.com/WordPress/gutenberg/pull/53876

**Comment by luisherranz** [comment]
The enhanced pagination of the Query block is ready for review:

- https://github.com/WordPress/gutenberg/pull/53812

**Comment by luisherranz** [comment]
There are still a couple of details regarding the movement of the form when a reply link is clicked, but otherwise, the enhanced comment form submission is ready for review:

- https://github.com/WordPress/gutenberg/pull/53737

**Comment by luisherranz** [comment]
We added support for `data-wp-slot` and `data-wp-fill` in the https://github.com/WordPress/gutenberg/pull/53737 PR, but we'll move it to its own PR:

- https://github.com/WordPress/gutenberg/pull/53737

**Comment by DAreRodz** [comment]
The PR for the new router with region-based, client-side navigation is merged:

- https://github.com/WordPress/gutenberg/pull/53733

**Comment by DAreRodz** [comment]
I just recently merged these PRs:
- https://github.com/WordPress/gutenberg/pull/54003
- https://github.com/WordPress/gutenberg/pull/53958

Just so you know, I left the JSDocs and documentation of the region-based router and the Slot and Fill directives for subsequent PRs. I'll add the task to the opening post.

**Comment by DAreRodz** [comment]
The enhanced pagination for the Query Block has been merged. 🚀 

- https://github.com/WordPress/gutenberg/pull/53812

**Comment by Poliuk** [comment]
I've opened a couple of requests for the Query Loop that need to be included in the original Tracking Issue (I don't have the permissions):

- https://github.com/WordPress/gutenberg/issues/54059 
- https://github.com/WordPress/gutenberg/issues/54060

**Comment by DAreRodz** [comment]
As mentioned in https://github.com/WordPress/gutenberg/pull/53737#issuecomment-1706624654, we have paused the development of the Comments Form submission with client-side navigation until we develop a robust system to handle new styles, scripts, and other assets.

- https://github.com/WordPress/gutenberg/pull/53737

**Comment by luisherranz** [comment]
I've opened a discussion to talk about a mechanism to detect blocks that are compatible with the Interactivity API so we can improve the reliability of the enhanced pagination of the Query block:

- https://github.com/WordPress/gutenberg/discussions/54194

**Comment by DAreRodz** [comment]
I just opened a PR for moving the "Enhanced pagination" toggle and updating its help message.

- https://github.com/WordPress/gutenberg/pull/54198

**Comment by DAreRodz** [comment]
This PR makes `navigate()` calls to be canceled when another starts. Ready for review.

- https://github.com/WordPress/gutenberg/pull/54201

**Comment by luisherranz** [comment]
David opened the PR to add a timeout to the client navigation:

- https://github.com/WordPress/gutenberg/pull/54474

**Comment by luisherranz** [comment]
A small update:

Last week, we focused on preparing a private version of the Interactivity API for the WP 6.4 release, which needs to ship Core blocks already using it. This version will remain private and subject to change, as explained in this tracking issue:

- https://github.com/WordPress/gutenberg/issues/54291

Even though there was a bit of overlap with this tracking issue, we opened a new one because there are some tasks unrelated to the Interactivity API itself, but once those are finished, we'll get back to this one to continue our work.

**Comment by luisherranz** [comment]
Now that WP 6.4 Beta 1 is out, we've closed [the WP 6.4 Tracking Issue](https://github.com/WordPress/gutenberg/issues/54291) and we got back to this one.

In particular, we've begun developing the new `store()` API, but our plan is first to test it in Woo to ensure that we don't find any significant issues there. If all goes well, we'll do the same in Gutenberg to see if there are any issues here. We'll use this Tracking Issue again once we move the development back to Gutenberg.

In the meantime, if you want to follow closely this development, please follow:

- https://github.com/woocommerce/woocommerce-blocks/issues/11065

**Comment by luisherranz** [comment]
We've completed all the tasks of the migration to the new `store()` API successfully in Woo.

- https://github.com/woocommerce/woocommerce-blocks/pull/11071

I've updated the description to include the next steps:

- Open a PR with the new API here in Gutenberg to make sure that the e2e tests that we have in Gutenberg pass.
- Migrate the Woo blocks to the new API to make sure that everything works fine and we don't find any blockers.
- Sync the code back to Gutenberg.
- Migrate the Core blocks to the new API to make sure that everything works fine and we don't fine any blockers.
- Add missing e2e tests.
- Migrate the documentation.
- Migrate the getting started template.

**Comment by luisherranz** [comment]
I've opened a couple of PRs to fix problems with `wp_unique_id` and the enhanced pagination: 

- https://github.com/WordPress/gutenberg/pull/55415
- https://github.com/WordPress/gutenberg/pull/55416

I've also added two more additional tasks to the list:

- Update the `<title>` on navigation.
- Turn off the enhanced pagination when certain blocks are inserted: Post Content, Template Part, Reusable block…

**Comment by DAreRodz** [comment]
I created a PR for testing the new `store()` API.

- https://github.com/WordPress/gutenberg/pull/55459

The runtime currently works fine, although we have probably broken the PHP part. I'll take a look later. 🙂

**Comment by DAreRodz** [comment]
I created a PR for disallowing core blocks that can render third-party blocks and be edited outside of the Query Loop context, e.g., Content, Template Part and Patterns.

- https://github.com/WordPress/gutenberg/pull/55539

**Comment by luisherranz** [comment]
@c4rl0sbr4v0 made this PR:

- https://github.com/WordPress/gutenberg/pull/55446

**Comment by cbravobernal** [comment]
PR with a fix for the position of some accesibility markup in the enhanced Query Loop has been merged:
 - https://github.com/WordPress/gutenberg/pull/55890

**Comment by luisherranz** [comment]
Most of the progress in the last weeks has been done in the migration to the new `store()` API PR:

- https://github.com/WordPress/gutenberg/pull/55459

That PR contains both the new runtime and the migration of the Core blocks because everything needs to be merged at the same time.

It's already almost finished, awaiting final details and reviews.

**Comment by DAreRodz** [comment]
The PR for migrating to the new `store()` API is complete. I've also marked all related tasks as done and added new ones for the remaining work.

- https://github.com/WordPress/gutenberg/pull/55459

**Comment by DAreRodz** [comment]
The `create-block-interactive-template` package was migrated to the new `store()` API here:

- https://github.com/WordPress/gutenberg/pull/56613

**Comment by luisherranz** [comment]
Now that the new `store()` API has been merged, let's see what we need to do before the 17.2 release next Wednesday:

- [x] Migrate the `create-block` template
  - https://github.com/WordPress/gutenberg/pull/56613
- [x] Migrate the docs
  - https://github.com/WordPress/gutenberg/issues/56640
  - https://github.com/WordPress/gutenberg/issues/56641
- [x] Migrate the proposal examples
  - [Interactivity API Proposal](https://make.wordpress.org/core/2023/03/30/proposal-the-interactivity-api-a-better-developer-experience-in-building-interactive-blocks/)
- [ ] Create a migration guide for the changelog
  - https://github.com/WordPress/gutenberg/discussions/52906

Does anything else need to be completed before the release?

**Comment by DAreRodz** [comment]
> Does anything else need to be completed before the release?

@luisherranz, maybe we should add and update JSDocs & TSDocs? 🤔 

- [x] Fix outdated docs: [`store()`](https://github.com/WordPress/gutenberg/blob/53ffc0b1ba97ca13355cfd81be3b85b58d7d9920/packages/interactivity/src/store.ts#L178-L218), [`directive()`](https://github.com/WordPress/gutenberg/blob/53ffc0b1ba97ca13355cfd81be3b85b58d7d9920/packages/interactivity/src/hooks.tsx#L108-L172)
- [x] Add missing docs: `getContext()`, `getElement()`, `navigate()`, `prefetch()`

**Comment by luisherranz** [comment]
Perfect, yes.

**Comment by DAreRodz** [comment]
I've recently created a PR to update TSDocs/JSDocs for the new `store()` API. Ready for review.

- https://github.com/WordPress/gutenberg/pull/56748

In addition, last week, I opened a PR draft with the ongoing work to implement `wp_initial_state`.

- https://github.com/WordPress/gutenberg/pull/56698

**Comment by luisherranz** [comment]
I've started the PR with the updated docs: 

- https://github.com/WordPress/gutenberg/pull/56764

**Comment by luisherranz** [comment]
Now that the new `store()` API has been merged, let's close this Tracking Issue as completed and start a new one to track the work to prepare the public 1.0 version.

- https://github.com/WordPress/gutenberg/issues/56803

**Comment by luisherranz** [comment]
I've published the migration guide for the new `store()` API:

- https://github.com/WordPress/gutenberg/discussions/52906#discussioncomment-7810998
